File: cdm/cpp/properties/SEScalarPressure.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarPressure.h"

const PressureUnit PressureUnit::Pa("Pa");
const PressureUnit PressureUnit::mmHg("mmHg");
const PressureUnit PressureUnit::cmH2O("cmH2O");
const PressureUnit PressureUnit::psi("psi");
const PressureUnit PressureUnit::atm("atm");

bool PressureUnit::IsValidUnit(const std::string& unit)
{
  if (Pa.GetString().compare(unit) == 0)
    return true;
  if (mmHg.GetString().compare(unit) == 0)
    return true;
  if (cmH2O.GetString().compare(unit) == 0)
    return true;
  if (psi.GetString().compare(unit) == 0)
    return true;
  if (atm.GetString().compare(unit) == 0)
    return true;
  return false;
}

const PressureUnit& PressureUnit::GetCompoundUnit(const std::string& unit)
{
  if (Pa.GetString().compare(unit) == 0)
    return Pa;
  if (mmHg.GetString().compare(unit) == 0)
    return mmHg;
  if (cmH2O.GetString().compare(unit) == 0)
    return cmH2O;
  if (psi.GetString().compare(unit) == 0)
    return psi;
  if (atm.GetString().compare(unit) == 0)
    return atm;
  std::stringstream err;
  err << unit << " is not a valid Pressure unit";
  throw CommonDataModelException(err.str());
}

void SEScalarPressure::Load(const cdm::ScalarPressureData& src, SEScalarPressure& dst)
{
  SEScalarPressure::Serialize(src, dst);
}
void SEScalarPressure::Serialize(const cdm::ScalarPressureData& src, SEScalarPressure& dst)
{
  SEScalarQuantity<PressureUnit>::Serialize(src.scalarpressure(), dst);
}

cdm::ScalarPressureData* SEScalarPressure::Unload(const SEScalarPressure& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarPressureData* dst = new cdm::ScalarPressureData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarPressure::Serialize(const SEScalarPressure& src, cdm::ScalarPressureData& dst)
{
  SEScalarQuantity<PressureUnit>::Serialize(src, *dst.mutable_scalarpressure());
}

File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarPressure.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarPressureData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.PressureUnit;

/**
 * @author abray
 * Pressure Scalar Class, will enforce that units are proper to Pressures
 */

public class SEScalarPressure extends SEScalar
{
  public SEScalarPressure()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarPressure(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarPressure(double value, PressureUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarPressureData src, SEScalarPressure dst)
  {
    SEScalar.load(src.getScalarPressure(),dst);
  }
  public static ScalarPressureData unload(SEScalarPressure src)
  {
    if(!src.isValid())
      return null;
    ScalarPressureData.Builder dst = ScalarPressureData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarPressure src, ScalarPressureData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarPressureBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, PressureUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(PressureUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(PressureUnit.validUnit(unit))
      return true;
    return false;
  }
}
